File: src/bin/day08.rs
use std::fs;

#[derive(Clone, Copy)]
enum Instr {
    NOp(i32),
    Acc(i32),
    Jmp(i32),
}

enum Exit {
    Terminated(i32),
    Seen(i32),
}

fn parse_line(ln: &str) -> Instr {
    let parts: Vec<_> = ln.split(' ').collect();
    assert_eq!(parts.len(), 2);
    let num = parts[1].parse().unwrap();
    match parts[0] {
        "nop" => Instr::NOp(num),
        "acc" => Instr::Acc(num),
        "jmp" => Instr::Jmp(num),
        _ => panic!("unknown instruction {instr}", instr = ln),
    }
}

fn run(program: &[Instr]) -> Exit {
    let mut acc = 0;
    let mut pc = 0;
    let mut seen = vec![false; program.len()];
    let term_pc = program.len();
    while !seen.get(pc).cloned().unwrap_or(false) && pc != term_pc {
        seen[pc] = true;
        match program[pc] {
            Instr::NOp(_) => pc += 1,
            Instr::Jmp(num) => pc = (pc as i32 + num) as usize,
            Instr::Acc(num) => { 
                acc += num; 
                pc += 1;
            },
        }
    }
    if pc == term_pc {
        Exit::Terminated(acc)
    } else{
        Exit::Seen(acc)
    }
}

fn part1(program: &[Instr]) -> i32 {
    if let Exit::Seen(num) = run(program) {
        num
    } else {
        panic!("unexpected exit code in part1");
    }
}


fn part2(program: &mut [Instr]) -> i32 {
    for i in 0..program.len() {
        let op = program[i];
        let new_op = match op {
            Instr::NOp(num) => Some(Instr::Jmp(num)),
            Instr::Jmp(num) => Some(Instr::NOp(num)),
            _ => None,
        };
        if !new_op.is_some() {
            continue;
        }
        program[i] = new_op.unwrap();
        if let Exit::Terminated(acc) = run(program) {
            return acc;
        }
        program[i] = op;
    }
    panic!("part2: couldn't make program terminate.")
}

fn main() {
    let contents = fs::read_to_string("inputs/day08.txt").expect("Something went wrong reading the file");

    let mut program: Vec<Instr> = contents.lines().map(parse_line).collect();

    println!("Part 1: {}", part1(&program));
    println!("Part 2: {}", part2(&mut program));
}

File: julia/day08.jl
# time ./julia day08.jl -> 255ms when using a single type to represent
# all operations:
# struct Instr
#  op :: Symbol
#  num :: Int64
# end
#
# Increases to 347ms when using separate types for
# each op (OpNop, OpJmp, OpAcc) and Union{..} for an instruction.

lines = open("../inputs/day08.txt") do f
    readlines(f)
end

# Instructions
struct OpNop
    num::Int64
end
struct OpJmp
    num::Int64
end
struct OpAcc
    num::Int64
end

isNop(i::OpNop) = true
isNop(_) = false

isJmp(i::OpJmp) = true
isJmp(_) = false

isAcc(i::OpAcc) = true
isAcc(_) = false

Instr = Union{OpNop,OpJmp,OpAcc}

# Possible terminations
struct Terminated
    acc::Int64
end

isTerminated(t::Terminated) = true
isTerminated(_) = false

struct Seen
    acc::Int64
end

# Parse

function parseLine(ln::String)::Instr
    parts = split(ln, " ")
    num = parse(Int64, parts[2])
    if parts[1] == "nop"
        return OpNop(num)
    elseif parts[1] == "jmp"
        return OpJmp(num)
    else
        @assert parts[1] == "acc"
        return OpAcc(num)
    end
end

function run(program::Vector{Instr})::Union{Terminated,Seen}
    acc = 0
    pc = 1
    seen = Set{Int64}()
    term_pc = length(program) + 1
    while pc ∉ seen && pc != term_pc
        push!(seen, pc)
        op = program[pc]
        if isNop(op)
            pc += 1
            continue
        elseif isJmp(op)
            pc += op.num
            continue
        else
            @assert isAcc(op)
            acc += op.num
            pc += 1
        end
    end
    if pc == term_pc
        return Terminated(acc)
    else
        return Seen(acc)
    end
end

part1(program) = run(program).acc

function part2(program)::Int64
    for i in 1:length(program)
        op = program[i]
        if isJmp(op)
            newOp = OpNop(op.num)
        elseif isNop(op)
            newOp = OpJmp(op.num)
        else
            continue
        end
        program[i] = newOp
        code = run(program)
        if isTerminated(code)
            return code.acc
        end
        program[i] = op
    end
end


function main()
    program::Vector{Instr} = parseLine.(lines)
    println("Part 1: ", part1(program))
    println("Part 2: ", part2(program))
end

main()